### ISSUE TITLE
Sometimes `testhost/CoreRun` crashes with `System.Diagnostics.Tracing`-related assertion failure

### ISSUE THREAD

**Issue by karakasa** [dotnet/runtime#92519]
### Description

`testhost/CoreRun` sometimes crashes for a blank console app.

### Reproduction Steps

1. Clone the latest repo and build it using `build -rc Debug`.
2. Use `testhost\CoreRun` to run a blank Console App (dll).

The crash doesn't always happen.

### Expected behavior

`CoreRun` runs without issues.

### Actual behavior

`CoreRun` crashes.

<details>
<summary>stack</summary>
<pre>
Process terminated. Assertion failed.
   at System.Diagnostics.Tracing.CounterGroup.OnEventSourceCommand(Object sender, EventCommandEventArgs e)
   at System.Diagnostics.Tracing.EventSource.add_EventCommandExecuted(EventHandler`1 value)
   at System.Diagnostics.Tracing.CounterGroup.RegisterCommandCallback()
   at System.Diagnostics.Tracing.CounterGroup..ctor(EventSource eventSource)
   at System.Diagnostics.Tracing.CounterGroup.GetCounterGroup(EventSource eventSource)
   at System.Diagnostics.Tracing.DiagnosticCounter.Publish()
   at System.Diagnostics.Tracing.PollingCounter..ctor(String name, EventSource eventSource, Func`1 metricProvider)
   at System.Diagnostics.Tracing.RuntimeEventSource.OnEventCommand(EventCommandEventArgs command)
   at System.Diagnostics.Tracing.EventSource.DoCommand(EventCommandEventArgs commandArgs)
   at System.Diagnostics.Tracing.EventSource.Initialize(Guid eventSourceGuid, String eventSourceName, String[] traits)
   at System.Diagnostics.Tracing.EventSource..ctor(Guid eventSourceGuid, String eventSourceName, EventSourceSettings settings, String[] traits)
   at System.Diagnostics.Tracing.RuntimeEventSource..ctor()
   at System.Diagnostics.Tracing.RuntimeEventSource.Initialize()
   at System.StartupHookProvider.ManagedStartup(Char* pDiagnosticStartupHooks)
</pre>
</details>

### Regression?

I never run into such issues in the past half a year. It started after I updated the local repo to the latest last afternoon.

### Known Workarounds

Rebooting Windows usually resolves the issue. But sometimes it occurs again after a period of usage.

### Configuration

Local .NET: 8 RC1
OS: Windows 10 x64 22H2

### Other information

Visual Studio debugger shows the assertion here fails.

https://github.com/dotnet/runtime/blob/0976df5dc5da790462d2f7cbe018ef82e4ec2af5/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs#L50

![image](https://github.com/dotnet/runtime/assets/12855165/c8d14154-48c5-4344-98b0-2fa2bb7d3263)

the unhandled `Update` seems coming from:

https://github.com/dotnet/runtime/blob/0976df5dc5da790462d2f7cbe018ef82e4ec2af5/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs#L810

**Comment by tarekgh** [comment]
@davmason could you have a look or get someone to triage this one? Thanks!

CC @@noahfalk

**Comment by karakasa** [comment]
hi @davmason 

Somehow this issue becomes very frequent today. I would like to perform a hack locally. Is it safe to just throw away the `Update` event in `CounterGroup.OnEventSourceCommand`?

**Comment by davmason** [comment]
Hi @karakasa, it should be safe to remove the Debug.Assert before you build. I'd like to understand the failure though, do you have the callstack of where it is crashing?

**Comment by karakasa** [comment]
> have the callstack of where it is crashing?

oh. I put it in a <details> block.

````
Process terminated. Assertion failed.
   at System.Diagnostics.Tracing.CounterGroup.OnEventSourceCommand(Object sender, EventCommandEventArgs e)
   at System.Diagnostics.Tracing.EventSource.add_EventCommandExecuted(EventHandler`1 value)
   at System.Diagnostics.Tracing.CounterGroup.RegisterCommandCallback()
   at System.Diagnostics.Tracing.CounterGroup..ctor(EventSource eventSource)
   at System.Diagnostics.Tracing.CounterGroup.GetCounterGroup(EventSource eventSource)
   at System.Diagnostics.Tracing.DiagnosticCounter.Publish()
   at System.Diagnostics.Tracing.PollingCounter..ctor(String name, EventSource eventSource, Func`1 metricProvider)
   at System.Diagnostics.Tracing.RuntimeEventSource.OnEventCommand(EventCommandEventArgs command)
   at System.Diagnostics.Tracing.EventSource.DoCommand(EventCommandEventArgs commandArgs)
   at System.Diagnostics.Tracing.EventSource.Initialize(Guid eventSourceGuid, String eventSourceName, String[] traits)
   at System.Diagnostics.Tracing.EventSource..ctor(Guid eventSourceGuid, String eventSourceName, EventSourceSettings settings, String[] traits)
   at System.Diagnostics.Tracing.RuntimeEventSource..ctor()
   at System.Diagnostics.Tracing.RuntimeEventSource.Initialize()
   at System.StartupHookProvider.ManagedStartup(Char* pDiagnosticStartupHooks)
````

**Comment by karakasa** [comment]
> have the callstack of where it is crashing?

ah. I put it in a `<details>` block.

![image](https://github.com/dotnet/runtime/assets/12855165/8d0cfe42-0f08-4648-bc90-a21fbeda00ec)


````
Process terminated. Assertion failed.
   at System.Diagnostics.Tracing.CounterGroup.OnEventSourceCommand(Object sender, EventCommandEventArgs e)
   at System.Diagnostics.Tracing.EventSource.add_EventCommandExecuted(EventHandler`1 value)
   at System.Diagnostics.Tracing.CounterGroup.RegisterCommandCallback()
   at System.Diagnostics.Tracing.CounterGroup..ctor(EventSource eventSource)
   at System.Diagnostics.Tracing.CounterGroup.GetCounterGroup(EventSource eventSource)
   at System.Diagnostics.Tracing.DiagnosticCounter.Publish()
   at System.Diagnostics.Tracing.PollingCounter..ctor(String name, EventSource eventSource, Func`1 metricProvider)
   at System.Diagnostics.Tracing.RuntimeEventSource.OnEventCommand(EventCommandEventArgs command)
   at System.Diagnostics.Tracing.EventSource.DoCommand(EventCommandEventArgs commandArgs)
   at System.Diagnostics.Tracing.EventSource.Initialize(Guid eventSourceGuid, String eventSourceName, String[] traits)
   at System.Diagnostics.Tracing.EventSource..ctor(Guid eventSourceGuid, String eventSourceName, EventSourceSettings settings, String[] traits)
   at System.Diagnostics.Tracing.RuntimeEventSource..ctor()
   at System.Diagnostics.Tracing.RuntimeEventSource.Initialize()
   at System.StartupHookProvider.ManagedStartup(Char* pDiagnosticStartupHooks)
````

**Comment by davmason** [comment]
Hi again @karakasa!

I spent some time trying to repro this but couldn't just by running a default console app. When you see this issue are you running any other performance collection tools? i.e. dotnet-trace, perfview, VS performance monitor, etc?

**Comment by karakasa** [comment]
> When you see this issue are you running any other performance collection tools? i.e. dotnet-trace, perfview, VS performance monitor, etc?

I was running benchmarks through `BenchmarkDotNet` and out of a sudden all tests using `CoreRun` started to fail. After that, I tried running a blank ConsoleApp with `CoreRun` from the terminal and got the assertion failure. I did have Visual Studio opened but I didn't run through it, nor was VS in the debugging mode.

I haven't experienced similar issues in the past 2 weeks so I assume something is wrong in the Windows environment. I would like to close the issue for now.
